packages/react-native-gesture-handler/src/v3/NativeDetector/utils.ts

```diff
@@ -0,0 +1,114 @@
+// This piece of magic traverses the gesture tree and populates `waitFor` and `simultaneousHandlers`
+// arrays for each gesture. It traverses the tree recursively using DFS.
+// `waitFor` and `simultaneousHandlers` are global data structures that will be populated into each gesture.
+// For `waitFor` we need array as order of the gestures matters.
+// For `simultaneousHandlers` we use Set as the order doesn't matter.
+
+import RNGestureHandlerModule from '../../RNGestureHandlerModule';
+import { isComposedGesture } from '../hooks/utils/relationUtils';
+import { ComposedGesture, ComposedGestureType, NativeGesture } from '../types';
+
+// The tree consists of ComposedGestures and NativeGestures. NativeGestures are always leaf nodes.
+export const traverseGestureRelations = (
+  node: NativeGesture | ComposedGesture,
+  simultaneousHandlers: Set<number>,
+  waitFor: number[] = []
+) => {
+  // If we are in the leaf node, we want to fill gesture relations arrays with current
+  // waitFor and simultaneousHandlers. We also want to configure relations on the native side.
+  if (!isComposedGesture(node)) {
+    node.gestureRelations.simultaneousHandlers.push(...simultaneousHandlers);
+    node.gestureRelations.waitFor.push(...waitFor);
+
+    RNGestureHandlerModule.configureRelations(node.tag, {
+      waitFor: node.gestureRelations.waitFor,
+      simultaneousHandlers: node.gestureRelations.simultaneousHandlers,
+      blocksHandlers: node.gestureRelations.blocksHandlers,
+    });
```

Comment:
Can this be done outside of this function? If not, can this be renamed? `traverseGestureRelations` in any way doesn't suggest that it updates the gestures on the native side.